### DISCUSSION TITLE
Why is only a single unique column supported for upserts?

### DISCUSSION THREAD

**Discussion by Oidlichtnwoada** [prisma/prisma#23887] [Q&A]
## Problem

Database upsert query criteria

Prisma Client uses a database upsert for an upsert query when the query meets the following criteria:

There are no nested queries in the upsert's create and update [options](https://www.prisma.io/docs/orm/prisma-client/queries/crud#options-7)
The query does not include a selection that uses a [nested read](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#nested-reads)
The query modifies only one model
-> There is only one unique field in the upsert's where option
The unique field in the where option and the unique field in the create option have the same value
If your query does not meet these criteria, then Prisma Client handles the upsert itself.

## Suggested solution

Support also multiple columns, it is supported for example in PostgreSQL:
```sql
INSERT INTO "Transaction" (
  "projectId",
  "storeId",
  "fromTimestamp",
  "toTimestamp",
  "productName",
  "productPrice",
  "productAmount",
  "invoiceAmount")
  VALUES 
	('asdfd2', 'hallo', '2017-07-28 11:42:42.846621+00', '2017-07-28 11:42:42.846621+00', 'office', 1276.5, 50.1, 1),
	('asdfd3', 'hallo', '2017-07-28 11:42:42.846621+00', '2017-07-28 11:42:42.846621+00', 'office', 1376.5, 50.1, 1),
	('asdfd4', 'hallo', '2017-07-28 11:42:42.846621+00', '2017-07-28 11:42:42.846621+00', 'office', 7136.5, 50.1, 1)
  ON CONFLICT ("projectId", "storeId", "fromTimestamp", "toTimestamp") DO UPDATE
    SET "productName" = excluded."productName",
        "productPrice" = excluded."productPrice",
		"productAmount" = excluded."productAmount",
		"invoiceAmount" = excluded."invoiceAmount";
```

**Comment by ludralph** [comment]
Hi @Oidlichtnwoada 👋 

Thank you for raising this question.

The Prisma Client's upsert operation currently supports only one unique field in the `where` option. This means that if you have multiple unique fields in the where clause, Prisma Client will not use a database upsert. This is one of the criteria for Prisma Client to use a database upsert. If this criteria is not met, Prisma Client handles the upsert itself.

In your suggested solution, you're using PostgreSQL's native SQL command to perform an upsert operation with multiple unique fields in the ON CONFLICT clause. While this is supported in PostgreSQL, it's not currently supported in Prisma Client's upsert operation. If you need to perform an upsert operation with multiple unique fields in the `where` clause, you may need to consider writing [raw queries](https://www.prisma.io/docs/orm/prisma-client/queries/raw-database-access/raw-queries#queryraw). To see this supported in the future, you can open a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) and document your usecase.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
